project3/coordinator/message_log.cpp

```diff
@@ -0,0 +1,37 @@
+/**
+ * @file Implementation of MessageLog.
+ */
+#include "message_log.h"
+namespace coordinator::message_log {
+
+bool MessageLog::Message::operator==(const Message &message) const {
+  return timestamp == message.timestamp && msg == message.msg &&
+         participant_id == message.participant_id;
+}
+
+size_t MessageLog::Hash::operator()(const Message &message) const {
+  // ensure uniqueness from this messages length and time stamp
+  return std::hash<int>()(message.msg.length() + message.timestamp);
+}
+
+void MessageLog::Insert(struct Message &msg) {
+  std::lock_guard<std::mutex> guard(mutex_);
+  coordinator_messages_.insert(msg);
+}
+void MessageLog::Clear() {
+  std::lock_guard<std::mutex> guard(mutex_);
+  coordinator_messages_.clear();
+}
+
+std::unordered_set<MessageLog::Message, MessageLog::Hash>
+MessageLog::GetMissedMessages(uint32_t reconnection_time) {
```

Comment:
Do we want this function to also automatically clear any messages that were received more than `td_` seconds ago?